packages/rx_bloc_cli/mason_templates/bricks/lib_mfa/__brick__/lib/lib_mfa/services/mfa_service.dart

```diff
@@ -0,0 +1,106 @@
+{{> licence.dart }}
+
+import 'dart:async';
+
+import 'package:rx_bloc/rx_bloc.dart';
+import 'package:rxdart/subjects.dart';
+
+import '../../lib_router/services/router_service.dart';
+import '../extensions/mfa_method_extensions.dart';
+import '../models/payload/mfa_payload_request.dart';
+import '../models/mfa_method.dart';
+import '../models/mfa_response.dart';
+import '../repositories/mfa_repository.dart';
+
+class MFAService {
+  MFAService(
+    this._mfaRepository,
+    this._routerService,
+  );
+
+  final RouterService _routerService;
+  final MFARepository _mfaRepository;
+
+  final BehaviorSubject<MFAResponse> _onResponse = BehaviorSubject();
+
+  /// Initiates the MFA process by the given [request].
+  ///
+  /// - [request] is the request body that contains the necessary user data to initiate the process.
+  /// Returns a [Stream] of [MFAResponse] that emits each step of the MFA process.
+  Stream<MFAResponse> authenticate({
+    required MFAPayloadRequest payload,
+  }) async* {
+    final response = await _mfaRepository.initiate(
+      action: payload.type,
+      request: payload,
+    );
+
+    yield* _executeAuthMethods(response);
+  }
+
+  /// Executes the MFA process by the given [response].
+  /// - [response] is the response that contains the necessary data to execute the MFA process.
+  /// Returns a [Stream] of [MFAResponse] that emits each step of the MFA process.
+  Stream<MFAResponse> _executeAuthMethods(
+    MFAResponse response,
+  ) async* {
+    MFAResponse? lastResponse = response;
+
+    while (true) {
+      if (lastResponse == null) {
+        // Complete the stream if the auth method returns null
+        break;
+      }
```

Comment:
I understand the concept of yielding values from a method from other languages, but in the context of dart, this is a rare case I've seen it being useful. In regards to the generator function yielding multiple values, that seems understandable to me. But in this particular case, I have a question.

What would happen if `lastResponse` is `null` at the start of the function? The while loop would break immediatelly and nothing would be yielded. In turn, nothing would be returned from the `_executeAuthMethods` method as well. What would then happen with the line `yield* _executeAuthMethods(response);` ? Would it just ignore the line or would it still "try" to return `null` or something?

Reply:
Yes the stream will complete, since we break the loop. It's equivalent to the `lastResponse.authMethod == MFAMethod.complete` check, which you can see in the same method. Let me know if this answer your question.